modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java

```diff
@@ -495,31 +517,28 @@ public void clear() {
 
     @Override
     public int hashCode() {
-        int hash = 7;
-        if (bits.length > 0) {
-            for (int n = 0; n < bits.length; n++) {
-                final long mask = bits[n];
-                hash = 71 * hash + (int)(mask ^ (mask >>> 32));
-            }
-        }
-        return hash;
+        // Note: overridden because equals is overridden; both equals and hashCode MUST
+        // respect the Set contract to interact correctly with sets of other types!
+        return super.hashCode();
     }
 
     @Override
     public boolean equals(Object obj) {
-
-        if (this == obj) {
+        // Note: overridden to provide a fast path; both equals and hashCode MUST respect
+        // the Set contract to interact correctly with sets of other types!
+        if (obj == this) {
             return true;
         }
-
-        if (obj == null || this.getClass() != obj.getClass()) {
-            return false;
+        if (obj instanceof BitSet<?> bitSet) {  // fast path if other is a BitSet
```

Comment:
I think we may need to add a 'component type' to BitSet, to make sure than BitSet<A> never intersects with BitSet<B>.

(Didn't we do it already, may be in some other PR?)

Reply:
We did, in #1076. However, I would like to not pull in all the `BitSet` fixes into this PR as this has nothing to do with `BitSet` and its many problems.

Reply:
Since it's an internal class, and both PseudoClassState and StyleClassSet are final, checking for class == class is probably sufficient.
Do you think we should add a warning to BitSet, it is not a problem since no-one will ever extend it?

Reply:
This is also solved in #1076:

     if (obj instanceof BitSet<?> bitSet && getElementType().equals(bitSet.getElementType()))

But since that requires pulling in even more changes (`getElementType`) I've left it there.

Relying on subclasses being final though seems like a bad idea for an abstract class, unless we seal it and use permits (or, just remove `BitSet` completely... it's purpose will be minimal if #1076 is accepted).

Reply:
that's my point.  I suppose we could use sealed class here, since javafx requires minimum java17?

Reply:
Do you think this needs to be changed as part of this PR?

Note that the `==` solution requires:
- sealing BitSet permitting only PseudoClassState and StyleClassSet
- adding a warning that if you ever unseal it, that you must re-evaluate the `equals` implementation
- adding a warning to PseudoClassState / StyleClassSet to re-evaluate `BitSet#equals` if you remove `final`

The `instanceof` solution allows arbitary subclasses, like any other Set implementation does (`AbstractSet#equals` for example use `instanceof Set`) and requires no further changes.

What I'm missing is the reasoning why you would want to change this. I realize that the other solution can also work, but that's not enough IMHO. For me the two options are practically equivalent, neither being truly better than the other, so then it is a matter of preference, and I slightly favor using `instanceof` here to avoid potential problems down the road.

Reply:
> Do you think this needs to be changed as part of this PR?

I'll answer this part of your question: No.

It seems out of scope for this particular bug fix, which is aimed at fixing an API and impl problem with `Match`, and should limit the changes to other classes to those needed to accomplish this.

It's great to point out existing weaknesses in, say, `BitSet`, but they should be addressed separately, after evaluating the need.